core/templates/pages/diagnostic-test-player-page/diagnostic-test.model.ts

```diff
@@ -0,0 +1,256 @@
+    this.setTopicIdToAncestorTopicIds();
+    this.setTopicIdToSuccessorTopicIds();
+  }
+
+  getEligibleTopicIds(): string[] {
+    return this._eligibleTopicIds;
+  }
+
+  getFailedTopicIds(): string[] {
+    return this._failedTopicIds;
+  }
+
+  getPassedTopicIds(): string[] {
+    return this._passedTopicIds;
+  }
+
+  getCurrentTopicId(): string {
+    return this._currentTopicId;
+  }
+
+  getAncestorsTopicIds(topicId: string): string[] {
+    return this._topicIdToAncestorTopicIds[topicId];
+  }
+
+  getSkippedTopicIds(): string[] {
+    return this._skippedTopicIds;
+  }
+
+  getTopicIdToAncestorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToAncestorTopicIds;
+  }
+
+  getTopicIdToSuccessorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToSuccessorTopicIds;
+  }
+
+  getSuccessorTopicIds(topicId: string): string[] {
+    return this._topicIdToSuccessorTopicIds[topicId];
+  }
+
+  incrementNumberOfAttemptedQuestions(incrementByValue: number): void {
+    this._totalNumberOfAttemptedQuestions += incrementByValue;
+  }
+
+  getNumberOfAttemptedQuestions(): number {
+    return this._totalNumberOfAttemptedQuestions;
+  }
+
+  setTopicIdToAncestorTopicIds(): void {
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      let ancestors: string[] = [];
+      let prerequisites: string[] = cloneDeep(
+        this._topicIdToPrerequisiteTopicIds[topicId]);
+
+      while (prerequisites.length > 0) {
+        let len = prerequisites.length;
+        let lastTopicId = prerequisites[len - 1];
+        prerequisites.splice(len - 1, 1);
+        if (ancestors.indexOf(lastTopicId) === -1) {
+          ancestors.push(lastTopicId);
+        }
+
+        prerequisites = prerequisites.concat(
+          this._topicIdToPrerequisiteTopicIds[lastTopicId]);
+      }
+      this._topicIdToAncestorTopicIds[topicId] = ancestors;
+    }
+  }
+
+  setTopicIdToSuccessorTopicIds(): void {
+    let topicIdToChildTopicId: TopicIdToRelatedTopicIds = {};
+
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      topicIdToChildTopicId[topicId] = [];
+    }
+
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      let prereq = this._topicIdToPrerequisiteTopicIds[topicId];
+      for (let prereqTopicId of prereq) {
+        topicIdToChildTopicId[prereqTopicId].push(topicId);
+      }
+    }
+
+    for (let topicId in topicIdToChildTopicId) {
+      let successors: string[] = [];
+      let children: string[] = topicIdToChildTopicId[topicId];
+
+      while (children.length > 0) {
+        let len = children.length;
+        let lastTopicId = children[len - 1];
+        children.splice(len - 1, 1);
+        if (successors.indexOf(lastTopicId) === -1) {
+          successors.push(lastTopicId);
+        }
+        children = children.concat(topicIdToChildTopicId[lastTopicId]);
+      }
+      this._topicIdToSuccessorTopicIds[topicId] = successors;
+    }
+  }
+
+  setCurrentTopicId(): void {
```

Comment:
Again, this isn't a "set". Perhaps "selectNextTopicIdToTest" or something like that?